client-protocol.md

```diff
@@ -3,64 +3,93 @@
 
-Each such message contains a "phase" string, and a hex-encoded binary "body".
+This document explains the format of these client-to-client messages. Each
+such message contains a "phase" string, and a hex-encoded binary "body":
+
+```json
+{
+  "phase": <string or number>,
+  "body": "<hex bytes>",
+}
+```
+
+From now on, all mentioned JSON messages will be from within the `body` field.
 
 Any phase which is purely numeric (`^\d+$`) is reserved for encrypted
 application data. The Rendezvous server may deliver these messages multiple
 times, or out-of-order, but the wormhole client will deliver the
 corresponding decrypted data to the application in strict numeric order. All
 other (non-numeric) phases are reserved for the Wormhole client itself.
 Clients will ignore any phase they do not recognize.
+The order of the wormhole phases are: `pake`, `version`, numerical.
+
+## Pake
 
 Immediately upon opening the mailbox, clients send the `pake` phase, which
 contains the binary SPAKE2 message (the one computed as `X+M*pw` or
-`Y+N*pw`).
+`Y+N*pw`):
+
+```json
+{
+  "pake_v1": <hex-encoded pake message>,
+}
+```
 
 Upon receiving their peer's `pake` phase, clients compute and remember the
-shared key. They derive the "verifier" (a hash of the shared key) and deliver
-it to the application by calling `got_verifier`: applications can display
+shared key. They derive a "verifier", which is a subkey of the shared key
+generated with `wormhole:verifier` as purpose): applications can display
 this to users who want additional assurance (by manually comparing the values
-from both sides: they ought to be identical). At this point clients also send
-the encrypted `version` phase, whose plaintext payload is a UTF-8-encoded
-JSON-encoded dictionary of metadata. This allows the two Wormhole instances
+from both sides: they ought to be identical). This protects against the threat
+where a man in the middle attacker correctly guesses the password.
+
+From this point on, all messages are encrypted using a NaCl `SecretBox` or some
+semantic equivalent. The key is derived from the shared key using the following
+purpose (as pseudocode-fstring): `f"wormhole:phase:{sha256(side)}{sha256(phase)}"`.
+The key derivation function is HKDF-SHA256, using the shared PAKE key as the secret.
+A random nonce is used, and nonce and ciphertext are concatenated. Their hex
+encoding is the content of the `body` field.
+
+## Version exchange
+
+In the `version` phase, the clients exchange information about themselves in
+order to do feature negotiation. 
+
+This allows the two Wormhole instances
 to signal their ability to do other things (like "dilate" the wormhole). The
 version data will also include an `app_versions` key which contains a
 dictionary of metadata provided by the application, allowing apps to perform
-similar negotiation.
+similar negotiation. At the moment, no keys other than `app_versions` are
+specified.
```

Comment:
If dilation is enabled in a client, they send a ``can-dilate`` key specifying the dilation versions they support. This is just `"1"` right now AFAICT so this will look like ``{"can-dilate": ["1"]}`` I believe.

Reply:
\*opens can of worms\*

I'd like to add a `abilities` list in order to promote features. This would allow easy handling by the implementations: simply take the union of ours and their set. This, however, somewhat of clashes with the concept of adding a `can-*ability*` field to each ability.

I think the dilation is not set in stone nor actively used in production, so my preferred approach would be to partially or fully enhance/replace `can-dilate` with the abilities list.

There a few other ways of going about this, but the main alternative is to add `can-*ability*` for other features as well.

Reply:
I just added a commit of I think how it could work

Reply:
Maybe you're right that it's best to just sort of ignore this for now .. Dilation isn't "official" (AFAIK) and so we could add appropriate words here along with some (future) concrete Dilation proposal. The `can-dilate` thing is just how it happens to work now (and the only implementation is the Python one still, AFAIK).